### DISCUSSION TITLE
Question: API URLs discovery

### DISCUSSION THREAD

**Discussion by franeko** [PiotrMachowski/Xiaomi-cloud-tokens-extractor#26] [Q&A] [answered]
Hi! First of all - great work on this extractor!

Could you tell me how you discovered API URLs for your requests?

I'm working on my own script to access data from BLE temperature and humidity sensors for HA integration. Those sensors appear on the device list but without additional data (although some devices do). I'm expecting something similar to `/v2/device/blt_get_beaconkey` but for sensors that could return temperature and humidity.

I've found [this website](http://miot-spec.org/miot-spec-v2/instance?type=urn:miot-spec-v2:device:light:0000A001:yeelight-lamp1:1) that seems to go into the right direction but I have no clue how to make use of it.

Any help will be greatly appreciated!

**Comment by PiotrMachowski** [comment]
Hi!

My work was based on two existing implementations:  [OpenHAB](https://github.com/openhab/openhab-addons/tree/main/bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud) and [valeCLOUDo](https://github.com/Xento/valeCLOUDo/blob/testing/lib/SmarthomeAPI.js), so I did't have to discover API URLs by myself.

BLE URL was a change proposed by @rezmus in [this issue](https://github.com/PiotrMachowski/Xiaomi-cloud-tokens-extractor/issues/15).

You can try to discover URLs by intercepting traffic from mobile app, e.g. using Wireshark.

**Reply by franeko** [reply]
@PiotrMachowski Thanks! I'll check those out

**Comment by rezmus** [comment]
@franeko you want to pull data from cloud for ble sensors (passed via ble gateways)?

**Reply by franeko** [reply]
Yes, that was my thought, I couldn't find any existing solution for it though

**Comment by rezmus** [comment, marked as answer]
```
/user/get_user_device_data
{"did":"blt.X","key":"4100","type":"prop","time_end":1620821594,"limit":10}
```

it will pull last 10 records of prop.4100 (temperature) for did blt.X starting from time_end timestamp. humidity is prop.4102, but you can fetch any prop/event stored for ble device this way. limit to 1 and ts = now for last record. from my experience if you want to fetch more history data (loop with time_end = oldest record from last request - 1) use more frequent requests with lower limit rather than bigger limit (it makes their db unhappy). not sure how much time they keep such data, but afair was able to pull data sent 2y ago.

**Reply by franeko** [reply]
Thanks! Sounds like exactly what I'm looking for, I'll check it out today

**Reply by franeko** [reply]
@rezmus thank you a lot!! This indeed works for my `miaomiaoce.sensor_ht.t2` device (square temp&humidity sensor) but doesn't for the `miaomiaoce.sensor_ht.t1` (with e-ink clock, like [this one](https://shopee.ph/Original-Xiaomi-Mi-Multifunctional-Digital-Clock-Electronic--i.143785093.2317498989)). 

Do you know if they have a separate list of property keys or maybe a separate API enpoint?

**Comment by rezmus** [comment]
model: miaomiaoce.sensor_ht.t1

conditions:
Higher than the specified temperature (prop.4100)
Lower than the specified temperature (prop.4100)
Higher than specified humidity (prop.4102)
Lower than specified humidity (prop.4102)

are you sure you have smart one and it's passing data to the cloud (you can see history in mijia app without active bt connection)?

**Reply by franeko** [reply]
Ok I think I know what the problem is - I have `t2` on the `DE` server and I have other devices acting as a BT gateway on the `DE` server - those work fine.

For `t1` I'm using the `CN` server and I don;t have any device acting as a BT gateway registered there so no data is uploaded to xiaomi. 

Does it make sense?

**Comment by rezmus** [comment]
no ble gw = no data, however there is some difference between ble devices using encrypted adv and these unencrypted. 1st can only work with gw on same region, because gw pulls ble key from the cloud to decrypt payload. 2nd may work with gw on many regions same time, because they don't need key to decrypt data. i don't remember if this sensor is encrypted or not, but if not it should also be able to upload data using gw on de region even if it's paired to cn region.

miaomiaoce.sensor_ht.t1 looks unencrypted (short ble key). did you try to pull data from de api?
